// search-engine-nodejs/lib/se-modules/Ask.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var request_promise_native_1 = __importDefault(require("request-promise-native"));
var cheerio_1 = __importDefault(require("cheerio"));
var fake_ua = require('fake-useragent');
var urlencode = require('urlencode');
var Aol = /** @class */ (function () {
    function Aol() {
        this.optionsDefault = {
            pageOfResult: 1,
            resolveWithFullResponse: true,
            jar: true,
            headers: {
                'User-Agent': fake_ua()
            },
        };
    }
    Aol.prototype.request = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var options, options, response, error_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        options = this.check(params);
                        options = __assign(__assign({}, this.optionsDefault), options);
                        return [4 /*yield*/, this.doRequest(options)];
                    case 1:
                        response = _a.sent();
                        return [3 /*break*/, 3];
                    case 2:
                        error_1 = _a.sent();
                        return [2 /*return*/, new Error("Google seems to suspect you as a robot, please try again later")];
                    case 3: return [2 /*return*/, response];
                }
            });
        });
    };
    Aol.prototype.doRequest = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            var response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.execRequest(options)];
                    case 1:
                        response = _a.sent();
                        return [2 /*return*/, response];
                }
            });
        });
    };
    Aol.prototype.check = function (options) {
        if (!options.qs)
            throw new Error("the qs attribute is required");
        if (!options.qs.q)
            throw new Error("the q attribute is required in qs");
        options.url = this.getUrl(options);
        options.qs.q = options.qs.q;
        return options;
    };
    Aol.prototype.execRequest = function (options) {
        return __awaiter(this, void 0, void 0, function () {
            var response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        options = this.pageOfResult(options);
                        return [4 /*yield*/, request_promise_native_1.default(options)];
                    case 1:
                        response = _a.sent();
                        return [4 /*yield*/, this.getLinks(response.body)];
                    case 2: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    Aol.prototype.getLinks = function (body) {
        return __awaiter(this, void 0, void 0, function () {
            var results;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.extractLinks(body)];
                    case 1:
                        results = _a.sent();
                        return [2 /*return*/, results.links];
                }
            });
        });
    };
    Aol.prototype.extractLinks = function (body) {
        var links = [];
        var $ = cheerio_1.default.load(body);
        $('body').find('.PartialSearchResults-item-title').map(function (i, data) {
            var title = $('body').find('.PartialSearchResults-item-title').map(function (i, title) { return $(title).text(); }).get(i).trim();
            var url = $('body').find('.PartialSearchResults-item-title').map(function (i, title) { return $(title).find('a').first().text(); }).get(i).trim();
            var description = $('body').find('.PartialSearchResults-item-abstract').map(function (i, desc) { return $(desc).text(); }).get(i).trim();
            links.push({ url: url, title: title, description: description });
        });
        return { links: links };
    };
    Aol.prototype.pageOfResult = function (options) {
        options.qs.page = parseInt(options.pageOfResult);
        return options;
    };
    Aol.prototype.getUrl = function (options) {
        return "https://" + (options.host || 'ask.com') + "/web";
    };
    return Aol;
}());
exports.default = new Aol();
